File: java/src/tarehart/rlbot/steps/wall/WallTouchStep.java
package tarehart.rlbot.steps.wall;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.intercept.InterceptCalculator;
import tarehart.rlbot.math.BallSlice;
import tarehart.rlbot.math.SpaceTime;
import tarehart.rlbot.math.VectorUtil;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.ArenaModel;
import tarehart.rlbot.physics.BallPath;
import tarehart.rlbot.physics.DistancePlot;
import tarehart.rlbot.carpredict.AccelerationModel;
import tarehart.rlbot.planning.GoalUtil;
import tarehart.rlbot.intercept.Intercept;
import tarehart.rlbot.planning.SteerUtil;
import tarehart.rlbot.steps.Step;
import tarehart.rlbot.time.Duration;

import java.awt.*;
import java.util.Optional;

import static java.util.Optional.empty;
import static java.util.Optional.of;
import static tarehart.rlbot.tuning.BotLog.println;

public class WallTouchStep implements Step {
    public static final double ACCEPTABLE_WALL_DISTANCE = ArenaModel.BALL_RADIUS + 5;
    public static final double WALL_DEPART_SPEED = 10;
    private static final double MIN_HEIGHT = 10;
    private Vector3 originalIntercept;

    private static boolean isBallOnWall(CarData car, SpaceTime ballPosition) {
        return ballPosition.space.z > MIN_HEIGHT && ArenaModel.getDistanceFromWall(ballPosition.space) <= ACCEPTABLE_WALL_DISTANCE;
    }

    private static boolean isBallOnWall(BallSlice ballPosition) {
        return ballPosition.space.z > MIN_HEIGHT && ArenaModel.getDistanceFromWall(ballPosition.space) <= ACCEPTABLE_WALL_DISTANCE;
    }

    public Optional<AgentOutput> getOutput(AgentInput input) {

        CarData car = input.getMyCarData();
        if (!ArenaModel.isCarOnWall(car)) {
            println("Failed to make the wall touch because the car is not on the wall", input.playerIndex);
            return empty();
        }


        BallPath ballPath = ArenaModel.predictBallPath(input);
        DistancePlot fullAcceleration = AccelerationModel.simulateAcceleration(car, Duration.ofSeconds(4), car.boost, 0);

        Optional<Intercept> interceptOpportunity = InterceptCalculator.getFilteredInterceptOpportunity(car, ballPath, fullAcceleration, new Vector3(), WallTouchStep::isBallOnWall);
        Optional<BallSlice> ballMotion = interceptOpportunity.flatMap(inter -> ballPath.getMotionAt(inter.getTime()));


        if (!ballMotion.isPresent()) {
            println("Failed to make the wall touch because we see no intercepts on the wall", input.playerIndex);
            return empty();
        }
        BallSlice motion = ballMotion.get();


        if (originalIntercept == null) {
            originalIntercept = motion.getSpace();
        } else {
            if (originalIntercept.distance(motion.getSpace()) > 20) {
                println("Failed to make the wall touch because the intercept changed", input.playerIndex);
                return empty(); // Failed to kick it soon enough, new stuff has happened.
            }
        }

        if (readyToJump(input, motion.toSpaceTime())) {
            println("Jumping for wall touch.", input.playerIndex);
            return of(new AgentOutput().withAcceleration(1).withJump());
        }

        return Optional.of(SteerUtil.steerTowardWallPosition(car, motion.space));
    }

    private static boolean readyToJump(AgentInput input, SpaceTime carPositionAtContact) {

        if (ArenaModel.getDistanceFromWall(carPositionAtContact.space) > ArenaModel.BALL_RADIUS + .5) {
            return false; // Really close to wall, no need to jump. Just chip it.
        }
        CarData car = input.getMyCarData();
        Vector3 toPosition = carPositionAtContact.space.minus(car.position);
        double correctionAngleRad = VectorUtil.getCorrectionAngle(car.orientation.noseVector, toPosition, car.orientation.roofVector);
        double secondsTillIntercept = Duration.between(input.time, carPositionAtContact.time).getSeconds();
        double wallDistanceAtIntercept = ArenaModel.getDistanceFromWall(carPositionAtContact.space);
        double tMinus = secondsTillIntercept - wallDistanceAtIntercept / WALL_DEPART_SPEED;
        boolean linedUp = Math.abs(correctionAngleRad) < Math.PI / 8;
        if (tMinus < 3) {
            println("Correction angle: " + correctionAngleRad, input.playerIndex);
        }

        return tMinus < 0.1 && tMinus > -.4 && linedUp;
    }

    @Override
    public boolean canInterrupt() {
        return true;
    }

    @Override
    public String getSituation() {
        return "Making a wall touch.";
    }

    public static boolean hasWallTouchOpportunity(AgentInput input, BallPath ballPath) {

        Optional<BallSlice> nearWallOption = ballPath.findSlice(WallTouchStep::isBallOnWall);
        if (nearWallOption.isPresent()) {
            BallSlice nearWall = nearWallOption.get();
            if (Duration.between(input.time, nearWall.getTime()).getSeconds() > 3) {
                return false; // Not on wall soon enough
            }

            Optional<BallSlice> ballLater = ballPath.getMotionAt(nearWall.getTime().plusSeconds(1));
            if (ballLater.isPresent()) {
                BallSlice motion = ballLater.get();
                if (ArenaModel.getDistanceFromWall(motion.getSpace()) > ACCEPTABLE_WALL_DISTANCE) {
                    return false;
                }
                Vector3 ownGoalCenter = GoalUtil.getOwnGoal(input.team).getCenter();
                return (motion.getSpace().distance(ownGoalCenter) > input.getMyCarData().position.distance(ownGoalCenter));
            }

        }

        return false;
    }

    @Override
    public void drawDebugInfo(Graphics2D graphics) {
        // Draw nothing.
    }
}

File: java/src/tarehart/rlbot/steps/wall/WallTouchStep.kt
package tarehart.rlbot.steps.wall

import tarehart.rlbot.AgentInput
import tarehart.rlbot.AgentOutput
import tarehart.rlbot.input.CarData
import tarehart.rlbot.intercept.InterceptCalculator
import tarehart.rlbot.math.BallSlice
import tarehart.rlbot.math.SpaceTime
import tarehart.rlbot.math.VectorUtil
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.physics.ArenaModel
import tarehart.rlbot.physics.BallPath
import tarehart.rlbot.carpredict.AccelerationModel
import tarehart.rlbot.planning.GoalUtil
import tarehart.rlbot.planning.SteerUtil
import tarehart.rlbot.steps.Step
import tarehart.rlbot.time.Duration

import java.awt.*

import tarehart.rlbot.tuning.BotLog.println

class WallTouchStep : Step {
    private var originalIntercept: Vector3? = null

    override val situation: String
        get() = "Making a wall touch."

    override fun getOutput(input: AgentInput): AgentOutput? {

        val car = input.myCarData
        if (!car.hasWheelContact) {
            println("Failed to make the wall touch because the car has no wheel contact", input.playerIndex)
            return null
        }


        val ballPath = ArenaModel.predictBallPath(input)
        val fullAcceleration = AccelerationModel.simulateAcceleration(car, Duration.ofSeconds(4.0), car.boost, 0.0)

        val interceptOpportunity = InterceptCalculator.getFilteredInterceptOpportunity(
                car, ballPath, fullAcceleration, Vector3(), { c: CarData, ballPosition: SpaceTime -> isBallOnWall(c, ballPosition) })
        val motion = interceptOpportunity?.let{ ballPath.getMotionAt(it.time) }


        if (motion == null) {
            println("Failed to make the wall touch because we see no intercepts on the wall", input.playerIndex)
            return null
        }


        originalIntercept?.let {
            if (it.distance(motion.space) > 20) {
                println("Failed to make the wall touch because the intercept changed", input.playerIndex)
                return null // Failed to kick it soon enough, new stuff has happened.
            }
        }

        val plane = ArenaModel.getNearestPlane(motion.space)
        if (plane.normal.z == 1.0) {
            println("Failed to make the wall touch because the ball is now close to the ground", input.playerIndex)
            return null
        }

        if (originalIntercept == null) {
            originalIntercept = motion.space
        }

        if (readyToJump(input, motion.toSpaceTime())) {
            println("Jumping for wall touch.", input.playerIndex)
            return AgentOutput().withThrottle(1.0).withJump()
        }

        return SteerUtil.steerTowardPositionAcrossSeam(car, motion.space)
    }

    override fun canInterrupt(): Boolean {
        return true
    }

    override fun drawDebugInfo(graphics: Graphics2D) {
        // Draw nothing.
    }

    companion object {
        val ACCEPTABLE_WALL_DISTANCE = (ArenaModel.BALL_RADIUS + 5).toDouble()
        val WALL_DEPART_SPEED = 10.0
        private val MIN_HEIGHT = 6.0

        private fun isBallOnWall(car: CarData, ballPosition: SpaceTime): Boolean {
            return ballPosition.space.z > MIN_HEIGHT && ArenaModel.getDistanceFromWall(ballPosition.space) <= ACCEPTABLE_WALL_DISTANCE
        }

        private fun isBallOnWall(ballPosition: BallSlice): Boolean {
            return ballPosition.space.z > MIN_HEIGHT && ArenaModel.getDistanceFromWall(ballPosition.space) <= ACCEPTABLE_WALL_DISTANCE
        }

        private fun readyToJump(input: AgentInput, carPositionAtContact: SpaceTime): Boolean {

            if (ArenaModel.getDistanceFromWall(carPositionAtContact.space) > ArenaModel.BALL_RADIUS + .7) {
                return false // Really close to wall, no need to jump. Just chip it.
            }
            val car = input.myCarData
            val toPosition = carPositionAtContact.space.minus(car.position)
            val correctionAngleRad = VectorUtil.getCorrectionAngle(car.orientation.noseVector, toPosition, car.orientation.roofVector)
            val secondsTillIntercept = Duration.between(input.time, carPositionAtContact.time).seconds
            val wallDistanceAtIntercept = ArenaModel.getDistanceFromWall(carPositionAtContact.space)
            val tMinus = secondsTillIntercept - wallDistanceAtIntercept / WALL_DEPART_SPEED
            val linedUp = Math.abs(correctionAngleRad) < Math.PI / 8
            if (tMinus < 3) {
                println("Correction angle: " + correctionAngleRad, input.playerIndex)
            }

            return tMinus < 0.1 && tMinus > -.4 && linedUp
        }

        fun hasWallTouchOpportunity(input: AgentInput, ballPath: BallPath): Boolean {

            val nearWallOption = ballPath.findSlice { ballPosition: BallSlice -> isBallOnWall(ballPosition) }
            if (nearWallOption != null) {
                val time = nearWallOption.time
                if (Duration.between(input.time, time).seconds > 3) {
                    return false // Not on wall soon enough
                }

                val ballLater = ballPath.getMotionAt(time.plusSeconds(1.0))
                if (ballLater != null) {
                    val (space) = ballLater
                    if (ArenaModel.getDistanceFromWall(space) > ACCEPTABLE_WALL_DISTANCE) {
                        return false
                    }
                    val ownGoalCenter = GoalUtil.getOwnGoal(input.team).center
                    return space.distance(ownGoalCenter) > input.myCarData.position.distance(ownGoalCenter)
                }

            }

            return false
        }
    }
}
